tests/bluetooth/host/keys/bt_keys_add_type/src/main.c

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/zephyr.h>
+#include <bluetooth/bluetooth.h>
+#include "testing_common_defs.h"
+#include <keys.h>
+#include "host_mocks/assert.h"
+#include "mocks/rpa.h"
+#include "mocks/keys_help_utils.h"
+
+/* This LUT contains different combinations of ID, Address and key type.
+ * Item in this list will be used to fill keys pool.
+ */
+static const struct id_addr_type testing_id_addr_type_lut[CONFIG_BT_MAX_PAIRED] = {
+
+	{ BT_ADDR_ID_1, BT_ADDR_LE_1, BT_KEYS_PERIPH_LTK },
+	{ BT_ADDR_ID_1, BT_RPA_ADDR_LE_1, BT_KEYS_PERIPH_LTK },
+	{ BT_ADDR_ID_1, BT_RPA_ADDR_LE_2, BT_KEYS_IRK },
+	{ BT_ADDR_ID_1, BT_ADDR_LE_3, BT_KEYS_IRK },
+
+	{ BT_ADDR_ID_2, BT_ADDR_LE_1, BT_KEYS_LTK },
+	{ BT_ADDR_ID_2, BT_RPA_ADDR_LE_2, BT_KEYS_IRK },
+	{ BT_ADDR_ID_2, BT_RPA_ADDR_LE_3, BT_KEYS_IRK },
+	{ BT_ADDR_ID_2, BT_ADDR_LE_2, BT_KEYS_LOCAL_CSRK },
+
+	{ BT_ADDR_ID_3, BT_ADDR_LE_1, BT_KEYS_REMOTE_CSRK },
+	{ BT_ADDR_ID_3, BT_ADDR_LE_2, BT_KEYS_LTK_P256 },
+
+	{ BT_ADDR_ID_4, BT_ADDR_LE_1, BT_KEYS_ALL }
+};
+
+/* Global iterator to iterate over the ID, Address and type LUT */
+static uint8_t params_it;
+
+/* Pointer to the current set of testing parameters */
+struct id_addr_type const *current_params_vector;
+
+/* This list will hold returned references while filling keys pool */
+static struct bt_keys *returned_keys_refs[CONFIG_BT_MAX_PAIRED];
+
+BUILD_ASSERT(ARRAY_SIZE(testing_id_addr_type_lut) == CONFIG_BT_MAX_PAIRED);
+BUILD_ASSERT(ARRAY_SIZE(testing_id_addr_type_lut) == ARRAY_SIZE(returned_keys_refs));
+
+/*
+ *  Set key type and verify the value
+ *
+ *  Constraints:
+ *   - Full keys pool list
+ *
+ *  Expected behaviour:
+ *   - The key type value is set correctly
+ */
+void test_set_type_value_correctly(void)
+{
+	struct bt_keys *key_ref;
+	int type = current_params_vector->type;
+
+	/* 'params_it' is incremented before running the test case */
+	key_ref = returned_keys_refs[params_it - 1];
```

Comment:
Isn't this returned_keys_refs equivalent to the result of bt_keys_get_addr() from the keys_fill_pool call? Isn't the value of elements in the returned_keys_ref array going to be indempotent between the keys_fill_pool() and the test iteration. If so, why not call bt_keys_get_addr here to add the entries as you use them, assert on failure and then directly use the result? That ensures that if tests are run out of order that entries are added in different orders and you better emulate any side-effects.

Reply:
Excellent remark